##### Imports and Utility Functions
Ensure the cell below is runs successfully to include all the helper utilities

In [41]:
%run workspaceutils

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 52, Finished, Available, Finished, True)

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


##### Prerequisites
Please read the disaster recovery guidance found in <a href="https://learn.microsoft.com/en-us/fabric/security/experience-specific-guidance">the documentation</a> to obtain a general understanding of how to recover lakehouse and warehouse data.
When the OneLake DR setting has been enabled at capacity level, lakehouse and warehouse data will be geo-replicated to the secondary (paired) region - which may not be accessible through the normal Fabric UI experience. Therefore, in a DR scenario, this data will need to be recovered into a corresponding (new) workspace in the DR region using the storage endpoints (abfs paths) as depicted in the image below. 
<!--<div style="margin: 0 auto; text-align: center; overflow: hidden;">
<div style="float: left;"> -->
<img src="https://github.com/hurtn/images/blob/main/reco_from_docs.png?raw=true" width="800"/>
<!--<small><b>Figure 1</b></small></div></div><br> -->

To use this recovery notebook please ensure:<p>&nbsp;</p>
1. You have imported and run the "01 - Run in Primary" notebook in the primary region.
<!-- <div style="margin: 0 auto; text-align: center; overflow: hidden;">
<div style="float: left;"> -->
<img src="https://github.com/hurtn/images/blob/main/before_recovery.png?raw=true" width="800"/>
<!-- <small><b>Figure 2</b></small></div></div><br> --><p>&nbsp;</p>
2. There is at least one capacity (C2) created in the DR region. 
<img src="https://github.com/hurtn/images/blob/main/starting_reco_stage1v2.png?raw=true" width="800"/>
<!--<small><b>Figure 3</b></small> </div></div><br> --><p>&nbsp;</p>
3. A recovery workspace (Reco2) has been created in the secondary region (C2.Reco2) which contains this notebook "02 - Run in DR" and attached to a default lakehouse (C2.Reco2.LH_BCDR).
<img src="https://github.com/hurtn/images/blob/main/starting_reco_stage2v2.png?raw=true" width="800"/>
<!-- <small><b>Figure 4</b></small> </div></div> -->
<p>&nbsp;</p>
4. Set the values for the notebook parameters below according to your item names. Sample names are shown below if using the naming scheme in the image above. Note: If passing these parameters in from a pipeline these defaults will be overridden:<p>&nbsp;</p>
<left>
p_bcdr_workspace_src -> Reco1 <br>
p_bcdr_workspace_tgt -> Reco2  <br>
p_bcdr_lakehouse_src -> LH_BCDR <br>
p_bcdr_lakehouse_tgt -> LH_BCDR <br>
p_secondary_ws_suffix eg '_SECONDARY' - specify a suffix which will be appended to each workspace created in the DR region. This is to ensure uniqueness of workspace names and is useful for identifying these worksspaces when testing in a non-DR scenario also.  <br>
p_recovered_object_suffix eg '_recovered'- specify a suffix that will be appended to each table name created in the recovery lakehouse C2.Reco2.LH_BCDR <br>
list_of_workspaces_to_recover eg ['WS1','WS2'] - specify a list of workspaces to specifically recover. This is useful for testing in a non-DR scenario also. Leave empty [''] for all workspaces.  <br>


In [42]:
# Specify the source and target BCDR workspaces and lakehouses
p_bcdr_workspace_src = 'Reco1' #'BCDR'
p_bcdr_workspace_tgt = 'Reco2' #'BCDR_DR'
p_bcdr_lakehouse_src = 'LH_BCDR' #'bcdrmeta'
p_bcdr_lakehouse_tgt = 'LH_BCDR' #'bcdrmeta'

# Specify the DR capacity ID or name. If left blank then the capacity of this workspace will be used
target_capacity = '51d3bcd5-9635-4ff5-98de-20f310a65ecb' 

# This variable adds a suffix to the name of the new workspaces created to ensure there are no naming conflicts with the original workspace name. 
# Ensure that you use a string that will gaurantee uniqueness rather than common terms which may be used by others in day to day activities.  
p_secondary_ws_suffix = '_DR'
p_recovered_object_suffix = '_recovered'

# Determines whether to add role assignments to the new workspaces. If you prefer to apply these at a later stage set the value to False. 
p_add_ws_role_assignments = True

# List parameters below need to be in the format of ['string1','string2',...'stringn']. Empty lists must be declared as []
# Specify the list of workspaces to recover, leave empty [] to recover all. For specific workspaces e.g. p_list_of_workspaces_to_recover = ['Workspace1','Workspace2'] 
p_list_of_workspaces_to_recover = ['W1','W2'] #['Prod1','Prod2'] #to specify exact workspaces
# Specify an exact list of workspaces to ignore e.g. p_ws_ignore_list = ['Microsoft Fabric Capacity Metrics 26/02/2024 16:15:42','AdminInsights']
p_ws_ignore_list = [] # add workspaces to this list to ignore them from the metadata extract process including git details
# Specify a list with wildcards using % e.g. to ignore anything with _DEV and _TEST as a suffix p_ws_ignore_like_list = ['%_DEV%','%_TEST%']  
p_ws_ignore_like_list  = [] #['%_DEV%','%_TEST%','%CLONE%']   #eg to specify exact ignore list ['BCDR']

# Boolean parameter to specify verbose informational messages. 
# Only set to True if additional logging information required, otherwise notebook may generate significant (print) messages.
p_logging_verbose = False

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 53, Finished, Available, Finished, False)

###### Check default lakehouse

In [43]:
if (notebookutils.runtime.context['defaultLakehouseId']==None):
    displayHTML('<div style="display: flex; align-items: flex-end;"><img style="float: left; margin-right: 10px;" src="https://github.com/hurtn/images/blob/main/stop.png?raw=true" width="50"><span><h4>Please set a default lakehouse before proceeding</span><img style="float: right; margin-left: 10px;" src="https://github.com/hurtn/images/blob/main/stop.png?raw=true" width="50"></div>')
    print('\n')
    raise noDefaultLakehouseException('No default lakehouse found. Please add a lakehouse to this notebook.')
else: 
    print('Default lakehouse is set to '+ notebookutils.runtime.context['defaultLakehouseName'] + '('+ notebookutils.runtime.context['defaultLakehouseId'] + ')')

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 54, Finished, Available, Finished, False)

Default lakehouse is set to LH_BCDR(873aa9bb-103f-4761-b810-96d7c1868c78)


###### Check target capacity

In [44]:
# Verify target capacity and status
if target_capacity != '':
    if _is_valid_uuid(target_capacity):
        target_capacity_id = target_capacity.lower()
    else: 
        # fetch the capacity ID
        target_capacity_id = labs.resolve_capacity_id(target_capacity)
else: # if not set then fetch capacity of this workspace
    target_capacity_id = labs.resolve_capacity_id()

# check capacity status
cap_status = get_capacity_status(target_capacity_id)
if cap_status == 'Inactive':
    raise ValueError(f"Status of capacity {target_capacity} is {cap_status}. Please resume the capacity and retry")
else:
    print(f"Status of capacity {target_capacity} is {cap_status}")


StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 55, Finished, Available, Finished, False)

Status of capacity 51d3bcd5-9635-4ff5-98de-20f310a65ecb is Active


##### Stage 4: Recover metadata tables from the BCDR workspace

In order for the recovery process to begin, it needs the metadata of the primary environment (created by running the Store DR Metadata notebook) such as workspaces. This data will be persisted as tables in the default lakehouse of this notebook.
<div>
<img src="https://github.com/hurtn/images/blob/main/starting_reco_stage2v3.png?raw=true" width="800"/>
</div>


In [45]:
# Gathers the list of recovers tables and source paths to be copied into the lakehouse associated with this notebook 

src_path = f'abfss://{p_bcdr_workspace_src}@onelake.dfs.fabric.microsoft.com/{p_bcdr_lakehouse_src}.Lakehouse'

table_list = get_lh_object_list(src_path)
print('The following tables will attempt to be recovered and persisted as tables in the default lakehouse of this notebook...')
display(table_list)

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 56, Finished, Available, Finished, False)

table name: capacities
delta log in names
capacities
table name: gitconnections
delta log in names
gitconnections
table name: items
delta log in names
items
table name: reports
delta log in names
reports
table name: workspaces
delta log in names
workspaces
table name: wsroleassignments
delta log in names
wsroleassignments
dbo
The following tables will attempt to be recovered and persisted as tables in the default lakehouse of this notebook...


SynapseWidget(Synapse.DataFrame, afdc236e-8dcd-4f25-a2e4-2b909a84a1f1)

In [ ]:
# wait for delta tables to be materialized in Reco1 from 01 - Run in DR notebook
time.sleep(60)

In [46]:
print('Restoring recovery tables...')
res = copy_lh_objects(table_list[table_list['type']=='table'],p_bcdr_workspace_src,p_bcdr_workspace_tgt,
                      p_bcdr_lakehouse_src,p_bcdr_lakehouse_tgt,p_recovered_object_suffix,False)
print('Done.')
display(res)

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 57, Finished, Available, Finished, False)

Restoring recovery tables...
Done.


SynapseWidget(Synapse.DataFrame, 79fd640c-7e0e-4121-9c4a-2c6b4ef254a4)

##### Stage 5: Recreate workspaces
Recover the workspaces that used to exist in the primary. The suffix parameter will be appended to the workspace name

<div>
<img src="https://github.com/hurtn/images/blob/main/Stage5_workspaces_created.png?raw=true" width="800"/>
</div>

In [47]:
from functools import reduce
from pyspark.sql.functions import col
import json

thisWsId = notebookutils.runtime.context["currentWorkspaceId"]

print("Creating secondary workspaces...")

# Discover available lakehouse tables/schemas
discovered_tables = infer_lakehouse_tables()

if p_logging_verbose:
    print("Discovered lakehouse tables:")
    print(json.dumps(discovered_tables, indent=2))

workspaces_recovered_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces" + p_recovered_object_suffix,
    preferred_schema=None
)

if p_logging_verbose:
    print("Using recovered workspaces path:", workspaces_recovered_path)

recovered_ws_df = (
    spark.read.format("delta")
    .load(workspaces_recovered_path)
    .select("Id", "Type", "Name", "Capacity_Id")
    .distinct()
)

recovered_ws_df = recovered_ws_df.filter(
    (col("Id") != thisWsId) &
    (col("Name") != p_bcdr_workspace_src)
)

if len(p_list_of_workspaces_to_recover) > 0:
    recovered_ws_df = recovered_ws_df.filter(
        col("Name").isin(p_list_of_workspaces_to_recover)
    )

if len(p_ws_ignore_list) > 0:
    recovered_ws_df = recovered_ws_df.filter(
        ~col("Name").isin(p_ws_ignore_list)
    )

if len(p_ws_ignore_like_list) > 0:
    conditions = [col("Name").like(f"%{p}%") for p in p_ws_ignore_like_list]
    recovered_ws_df = recovered_ws_df.filter(
        ~reduce(lambda a, b: a | b, conditions)
    )

recovered_ws_df.show(truncate=False)

created_workspace_names = []

for i in recovered_ws_df.collect():
    if i["Type"] == "Workspace":
        new_workspace_name = i["Name"] + p_secondary_ws_suffix

        try:
            if p_logging_verbose:
                print(
                    f"Creating workspace: {new_workspace_name} "
                    f"in target capacity {target_capacity_id}..."
                )

            response = fabric.create_workspace(
                new_workspace_name,
                target_capacity_id
            )

            if p_logging_verbose:
                print(f"Created workspace with response: {response}")

            created_workspace_names.append(new_workspace_name)

        except Exception as error:
            print(
                f"Failed to recreate workspace {new_workspace_name} "
                f"with capacity ID ({i['Capacity_Id']}) due to: {error}"
            )

print("Reloading workspace metadata table...")
saveWorkspaceMeta()
print("Workspace creation phase done.")

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 58, Finished, Available, Finished, False)

Creating secondary workspaces...
+------------------------------------+---------+----+------------------------------------+
|Id                                  |Type     |Name|Capacity_Id                         |
+------------------------------------+---------+----+------------------------------------+
|b11cf9d8-5334-46d7-9159-9145c76f3ef1|Workspace|W1  |ddf79b3c-8a34-4da3-8ba3-41113fc473b1|
|24229fbf-ab93-4a2d-b906-0d08c3d1917b|Workspace|W2  |ddf79b3c-8a34-4da3-8ba3-41113fc473b1|
+------------------------------------+---------+----+------------------------------------+

Reloading workspace metadata table...
57 records saved to the workspaces table.
Workspace creation phase done.


#### Stage 6: Connect Git and Initialise


Iterate through workspaces and connect them to the same branch they were connected to in primary.
Then intialise and update from git to start synchronising items from the repo.
<div>
<img src="https://github.com/hurtn/images/blob/main/Stage6_gitconnected.png?raw=true" width="800"/>
</div>

In [48]:
import json
import time
from pyspark.sql.functions import col, regexp_replace, lit

print("Reconnecting workspaces to Git...")

# Discover available lakehouse tables/schemas
discovered_tables = infer_lakehouse_tables()

if p_logging_verbose:
    print("Discovered lakehouse tables:")
    print(json.dumps(discovered_tables, indent=2))

gitconnections_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="gitconnections" + p_recovered_object_suffix,
    preferred_schema=None
)

workspaces_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces",
    preferred_schema=None
)

if p_logging_verbose:
    print("Using gitconnections path:", gitconnections_path)
    print("Using workspaces path:", workspaces_path)

git_df = (
    spark.read.format("delta")
    .load(gitconnections_path)
    .select(
        "gitConnectionState",
        "gitProviderDetails",
        "Workspace_Name",
        "ownerName",
        "repositoryName",
        "gitProviderType"
    )
)

# Refresh current workspaces after creation
wks_df = (
    spark.read.format("delta")
    .load(workspaces_path)
    .select("Name", "Id")
)

joined_df = (
    git_df.alias("gt")
    .join(
        wks_df.alias("wks"),
        col("gt.Workspace_Name") == regexp_replace(col("wks.Name"), lit(p_secondary_ws_suffix), ""),
        "inner"
    )
    .filter(col("gt.gitConnectionState") == "ConnectedAndInitialized")
    .filter(col("wks.Name").endswith(p_secondary_ws_suffix))
    .filter(col("wks.Id") != thisWsId)
)

if len(created_workspace_names) > 0:
    joined_df = joined_df.filter(
        col("wks.Name").isin(created_workspace_names)
    )

if len(p_list_of_workspaces_to_recover) > 0:
    joined_df = joined_df.filter(
        col("gt.Workspace_Name").isin(p_list_of_workspaces_to_recover)
    )

if len(p_ws_ignore_list) > 0:
    joined_df = joined_df.filter(
        ~col("gt.Workspace_Name").isin(p_ws_ignore_list)
    )

if len(p_ws_ignore_like_list) > 0:
    for notlike in p_ws_ignore_like_list:
        joined_df = joined_df.filter(
            ~col("gt.Workspace_Name").like(f"%{notlike}%")
        )

git_rows = (
    joined_df.select(
        col("gt.gitConnectionState").alias("gitConnectionState"),
        col("gt.gitProviderDetails").alias("gitProviderDetails"),
        col("gt.Workspace_Name").alias("Source_Workspace_Name"),
        col("wks.Name").alias("Workspace_Name"),
        col("wks.Id").alias("Workspace_ID"),
        col("gt.ownerName").alias("ownerName"),
        col("gt.repositoryName").alias("repositoryName"),
        col("gt.gitProviderType").alias("providerType")
    )
    .collect()
)

success = True

for i in git_rows:
    workspace_id = i["Workspace_ID"]
    workspace_name = i["Workspace_Name"]
    source_workspace_name = i["Source_Workspace_Name"]
    owner_name = i["ownerName"]
    repo_name = i["repositoryName"]
    provider_type = i["providerType"]
    row_success = True

    git_provider_details = (
        json.loads(i["gitProviderDetails"])
        if isinstance(i["gitProviderDetails"], str)
        else i["gitProviderDetails"]
    )

    print(repo_name)
    print(owner_name)
    print(provider_type)

    connection_id = get_connection_id_for_repo(
        repo_name,
        owner_name,
        provider_type
    )

    try:
        print(f"Workspace {workspace_name} using connection {connection_id}")

        # CONNECT FIRST
        url = f"/v1/workspaces/{workspace_id}/git/connect"
        payload = {
            "gitProviderDetails": git_provider_details,
            "myGitCredentials": {
                "source": "ConfiguredConnection",
                "connectionId": connection_id
            }
        }

        print(f"Attempting to connect workspace {workspace_name}")
        response = client.post(url, json=payload)

        if p_logging_verbose:
            print(str(response.status_code) + " " + response.text)

    except Exception as error:
        error_message = str(error)

        # already connected is not fatal
        if "already connected to Git" in error_message:
            print(f"Workspace {workspace_name} already connected to Git, continuing...")
            row_success = True
        else:
            row_success = False
            success = False
            print(
                f"Couldn't connect git to workspace {workspace_name} "
                f"({workspace_id}). Error: {error_message}"
            )

    if row_success:
        try:
            url = f"/v1/workspaces/{workspace_id}/git/initializeConnection"
            payload = {"initializationStrategy": "PreferRemote"}

            print(f"Attempting to initialize git connection for workspace {workspace_name}")
            response = client.post(url, json=payload)

            if p_logging_verbose:
                print(str(response.status_code) + " " + response.text)

            commithash = response.json().get("remoteCommitHash", "")

            if commithash != "":
                url = f"/v1/workspaces/{workspace_id}/git/updateFromGit"
                payload = {
                    "remoteCommitHash": commithash,
                    "conflictResolution": {
                        "conflictResolutionType": "Workspace",
                        "conflictResolutionPolicy": "PreferWorkspace"
                    },
                    "options": {
                        "allowOverrideItems": True
                    }
                }

                response = client.post(url, json=payload)

                if response.status_code == 202:
                    print("Successfully started a sync, LRO in progress...")
                    location_url = response.headers.get("Location")
                    print(f"Polling URL to track operation status is {location_url}")
                    time.sleep(15)
                    long_running_operation_polling(location_url, 15)

        except Exception as error:
            row_success = False
            success = False
            print(
                f"Couldn't initialize git for workspace {workspace_name} "
                f"({workspace_id}). Error: {error}"
            )

if success:
    print("Done")
else:
    print("Completed with errors.")

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 59, Finished, Available, Finished, False)

Reconnecting workspaces to Git...
FabricBCDRDemo-W1
iamjenetzler
GitHub
Workspace W1_DR using connection 831696e9-ef5b-4025-bc81-4d2f6f9cb43e
Attempting to connect workspace W1_DR
Attempting to initialize git connection for workspace W1_DR
Successfully started a sync, LRO in progress...
Polling URL to track operation status is https://wabi-west-us3-a-primary-redirect.analysis.windows.net/v1/operations/bcb79be5-044c-4a93-98f1-9ca32fcb632f
Polling long running operation ID https://wabi-west-us3-a-primary-redirect.analysis.windows.net/v1/operations/bcb79be5-044c-4a93-98f1-9ca32fcb632f has been started with a retry-after time of 15 seconds.
operation state = {'status': 'Running', 'createdTimeUtc': '2026-03-19T20:35:19.2154192', 'lastUpdatedTimeUtc': '2026-03-19T20:35:27.2705904', 'percentComplete': None, 'error': None}
Long running operation status: Running
operation state = {'status': 'Running', 'createdTimeUtc': '2026-03-19T20:35:19.2154192', 'lastUpdatedTimeUtc': '2026-03-19T20:35:27.27

##### After git sync completes then capture environment metadata (items and reports)  
#TODO: Need to write a method to check the sync is complete before attempting to extract the items. For now you will need to wait until the sync has complete and all workspaces have been updated from git. For now adding a sleep into the process to make sure it waits a minute for the sync to complete.

In [50]:
print(created_workspace_names)

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 61, Finished, Available, Finished, False)

['W1_DR', 'W2_DR']


In [51]:
import time
print('Gathering metadata about reports and items... ')

# variable that holds the names of the _DR workspaces we created
p_created_workspace_names = created_workspace_names

saveItemMeta(verbose_logging=p_logging_verbose, ws_ignore_list=p_ws_ignore_list,ws_ignore_like_list=p_ws_ignore_like_list,list_of_workspaces_to_recover=p_list_of_workspaces_to_recover,created_workspace_names=p_created_workspace_names)
saveReportMeta(verbose_logging=p_logging_verbose,only_secondary=True,ws_ignore_list=p_ws_ignore_list,ws_ignore_like_list=p_ws_ignore_like_list)

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 62, Finished, Available, Finished, False)

Gathering metadata about reports and items... 
57 records saved to the workspaces table.
20 records saved to the items table.
59 records saved to the reports table.


##### Stage 7: Recover lakehouse data (files and tables) to corresponding recovered workspaces

Copy lakehouse data (files and tables) from source abfs path to target lakehouse abfs path

<div>
<img src="https://github.com/hurtn/images/blob/main/lh_reocvery.png?raw=true" width="800"/>
</div>

In [52]:
from pyspark.sql import functions as F
from pyspark.sql.functions import col, lit, regexp_replace
import json

thisWsId = notebookutils.runtime.context["currentWorkspaceId"]  # failsafe: avoid overwriting objects in this workspace

discovered_tables = infer_lakehouse_tables()

if p_logging_verbose:
    print("Discovered lakehouse tables:")
    print(json.dumps(discovered_tables, indent=2))

# Resolve Delta table paths in a schema-agnostic way
items_recovered_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="items_recovered",
    preferred_schema=None
)

workspaces_recovered_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces_recovered",
    preferred_schema=None
)

capacities_recovered_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="capacities_recovered",
    preferred_schema=None
)

workspaces_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces",
    preferred_schema=None
)

items_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="items",
    preferred_schema=None
)

if p_logging_verbose:
    print("Using items_recovered path:", items_recovered_path)
    print("Using workspaces_recovered path:", workspaces_recovered_path)
    print("Using capacities_recovered path:", capacities_recovered_path)
    print("Using workspaces path:", workspaces_path)
    print("Using items path:", items_path)

# Load all source tables from Delta paths
items_recovered_df = (
    spark.read.format("delta")
    .load(items_recovered_path)
    .select("WorkspaceId", "Type", "DisplayName", "Id")
)
# display(items_recovered_df)

workspaces_recovered_df = (
    spark.read.format("delta")
    .load(workspaces_recovered_path)
    .select("Id", "Name", "Capacity_Id")
)
# display(workspaces_recovered_df)

capacities_recovered_df = (
    spark.read.format("delta")
    .load(capacities_recovered_path)
    .select("Id", "Display_Name")
)
# display(capacities_recovered_df)

workspaces_df = (
    spark.read.format("delta")
    .load(workspaces_path)
    .select("Id", "Name")
)
# display(workspaces_df)

items_df = (
    spark.read.format("delta")
    .load(items_path)
    .select("WorkspaceId", "Type", "DisplayName", "Id")
)
# display(items_df)

# Build equivalent joins
df_recovery_items = (
    items_recovered_df.alias("ir")
    .join(
        workspaces_recovered_df.alias("wr"),
        col("wr.Id") == col("ir.WorkspaceId"),
        "inner"
    )
    .join(
        capacities_recovered_df.alias("cr"),
        col("wr.Capacity_Id") == col("cr.Id"),
        "inner"
    )
    .join(
        workspaces_df.alias("wks"),
        col("wr.Name") == regexp_replace(col("wks.Name"), lit(p_secondary_ws_suffix), ""),
        "inner"
    )
    .join(
        items_df.alias("it"),
        (col("ir.DisplayName") == col("it.DisplayName")) &
        (col("it.WorkspaceId") == col("wks.Id")) &
        (col("it.Type") == col("ir.Type")),
        "inner"
    )
    .filter(col("ir.Type").isin("Lakehouse", "Warehouse"))
    .filter(col("wks.Name").endswith(p_secondary_ws_suffix))
    .filter(col("wks.Id") != thisWsId)
)

if len(p_list_of_workspaces_to_recover) > 0:
    df_recovery_items = df_recovery_items.filter(
        col("wr.Name").isin(p_list_of_workspaces_to_recover)
    )

if len(p_ws_ignore_list) > 0:
    df_recovery_items = df_recovery_items.filter(
        ~col("wr.Name").isin(p_ws_ignore_list)
    )

if len(p_ws_ignore_like_list) > 0:
    for notlike in p_ws_ignore_like_list:
        df_recovery_items = df_recovery_items.filter(
            ~col("wr.Name").like(f"%{notlike}%")
        )

# Final projection to match your SQL output
df_recovery_items = (
    df_recovery_items.select(
        col("wr.Name").alias("primary_ws_name"),
        col("ir.Type").alias("primary_type"),
        col("ir.DisplayName").alias("primary_name"),
        col("wr.Id").alias("primary_ws_id"),
        col("ir.Id").alias("primary_id"),
        col("wks.Id").alias("secondary_ws_id"),
        col("wks.Name").alias("secondary_ws_name"),
        col("it.Id").alias("secondary_id"),
        col("it.DisplayName").alias("secondary_name"),
        col("cr.Display_Name").alias("capacity_name")
    )
    .orderBy("primary_type", "primary_ws_name")
)
# display(df_recovery_items)

if df_recovery_items.count() > 0:
    print("The subsequent notebook cells attempt to recover lakehouse and warehouse data for the following items...")
else:
    print("No lakehouses and warehouses found to recover.")

lakehousedf = df_recovery_items.filter(col("primary_type") == "Lakehouse").collect()
warehousedf = df_recovery_items.filter(col("primary_type") == "Warehouse").collect()

display(
    df_recovery_items.select(
        "primary_ws_name",
        "primary_type",
        "primary_name",
        "secondary_ws_name",
        "secondary_name",
        "secondary_ws_id",
        "capacity_name"
    )
)

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 63, Finished, Available, Finished, False)

The subsequent notebook cells attempt to recover lakehouse and warehouse data for the following items...


SynapseWidget(Synapse.DataFrame, be176274-b430-49a2-8cee-02acbe55ca59)

Traceback (most recent call last) Cell In[37], 
line 30     
28 table_list = get_wh_object_list(schema_list['name'],src_path)     
29 #recover tables---> 
30 copy_lh_objects(table_list[table_list['type']=='table'],i['primary_ws_id'], i['secondary_ws_id'] ,i['primary_id'],i['secondary_id'],p_recovered_object_suffix,False,True)     
31 table_list = get_lh_object_list(src_path,['Files'])     
32 #recover files File ~/cluster-env/trident_env/lib/python3.11/site-packages/pandas/core/frame.py:3893, in DataFrame.__getitem__(self, key)   
3891 if selcolumns.nlevels > 1:   3892     return self._getitem_multilevel(key)-> 3893 indexer = selcolumns.get_loc(key)   3894 if is_integer(indexer):   3895     indexer = [indexer] File ~/cluster-env/trident_env/lib/python3.11/site-packages/pandas/core/indexes/range.py:418, in RangeIndex.get_loc(self, key)    416         raise KeyError(key) from err    417 if isinstance(key, Hashable):--> 418     raise KeyError(key)    419 self._check_indexing_error(key)    420 raise KeyError(key) KeyError: 'type' 


In [53]:
print('The following lakehouse(s) will attempt to be recovered... ')
display(lakehousedf)

for idx,i in enumerate(lakehousedf):
    # Define the full abfss source path of the primary BCDR workspace and lakehouse 
    src_path = f'abfss://'+i['primary_ws_id']+'@onelake.dfs.fabric.microsoft.com/'+i['primary_id']
    if p_logging_verbose:
        print('Attempting to recover items for workspace: ' + i['primary_ws_name'] + ', lakehouse: ' + i['primary_name'] + ' into target workspace ' + i['secondary_ws_name'] + ' lakehouse ' + i['secondary_name'])
    if isSchemaEnabledLakehouse(i['primary_ws_id'],i['primary_id']):
        # schema_list = get_lh_object_list(src_path,['Tables'])
        # table_list = get_wh_object_list(schema_list['name']) 

        lh_object_list = get_lh_object_list(src_path, data_types=["Tables", "Files"])

        table_list = lh_object_list[lh_object_list["type"] == "table"]
        file_list = lh_object_list[lh_object_list["type"] == "file"]

        # ,src_path)
        if len(table_list)>0:         #check there are tables present in the lakehouse before attempt to copy
            #recover tables
            copy_lh_objects(table_list[table_list['type']=='table'],i['primary_ws_id'], i['secondary_ws_id'] ,i['primary_id'],i['secondary_id'],p_recovered_object_suffix,False,True)
            table_list = get_lh_object_list(src_path,['Files'])
            #recover files
            copy_lh_objects(table_list[table_list['type']=='file'],i['primary_ws_id'], i['secondary_ws_id'] ,i['primary_id'],i['secondary_id'],p_recovered_object_suffix,False,True)
        else:
            print(f'No tables found in lakehouse {i["primary_name"]}')
    else:
        table_list = get_lh_object_list(src_path)
        copy_lh_objects(table_list[table_list['type']=='file'],i['primary_ws_id'], i['secondary_ws_id'] ,i['primary_id'],i['secondary_id'],p_recovered_object_suffix,False,True)
        #recover tables
        copy_lh_objects(table_list[table_list['type']=='table'],i['primary_ws_id'], i['secondary_ws_id'] ,i['primary_id'],i['secondary_id'],p_recovered_object_suffix,False,True)

print('Done')

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 64, Finished, Available, Finished, False)

The following lakehouse(s) will attempt to be recovered... 


SynapseWidget(Synapse.DataFrame, b84e7afe-25fa-48f6-ade3-dfc00a39d5b2)

table name: dimension_city
delta log in names
dimension_city
table name: dimension_customer
delta log in names
dimension_customer
table name: dimension_date
delta log in names
dimension_date
table name: dimension_employee
delta log in names
dimension_employee
table name: dimension_stock_item
delta log in names
dimension_stock_item
table name: fact_sale
delta log in names
fact_sale
dbo
dbo
table name: green_tripdata_2017
delta log in names
green_tripdata_2017
year_2017
table name: green_tripdata_2018
delta log in names
green_tripdata_2018
year_2018
table name: green_tripdata_2019
delta log in names
green_tripdata_2019
year_2019
table name: green_tripdata_2020
delta log in names
green_tripdata_2020
year_2020
table name: green_tripdata_2021
delta log in names
green_tripdata_2021
year_2021
table name: green_tripdata_2022
delta log in names
green_tripdata_2022
year_2022
Done


##### Stage 8: Prepare warehouse recovery
###### Important: 
This process creates a staging lakehouse to store shortcuts which point back to the DR copy of the tables (actually delta folders) - in testing this will point back to folders in the primary region but in a true DR scenario where Microsoft has failed over the OneLake endpoints, they will point back to folders in the secondary (paired) storage region.
A parameterised pipeline is injected into each target workspace which will load target tables from these shortcuts.
This staging lakehouse and pipeline can be deleted manually after the datawarehouse tables have been successfully recovered.
<div>
<img src="https://github.com/hurtn/images/blob/main/wh_recovery.png?raw=true" width="1000"/>
</div>

In [54]:
# iterate through all the data warehouses to recover
for idx, i in enumerate(warehousedf):
    if p_logging_verbose:
        print(
            'Setting up for recovery of warehouse '
            + i['primary_ws_name'] + '.' + i['primary_name']
            + ' into ' + i['secondary_ws_name'] + '.' + i['secondary_name']
        )

    src_path = f"abfss://{i['primary_ws_id']}@onelake.dfs.fabric.microsoft.com/{i['primary_id']}"
    tgt_path = f"abfss://{i['secondary_ws_id']}@onelake.dfs.fabric.microsoft.com/{i['secondary_id']}"

    lh_object_list = get_lh_object_list(src_path, data_types=["Tables", "Files"])
    print("lh_object_list:")
    display(lh_object_list)

    # extract the list of schemas per data
    schema_list = lh_object_list[lh_object_list["type"] == "table"]
    print("schema list:")
    display(schema_list)

    # extract a list of warehouse objects per schema and store in a list
    table_list = lh_object_list[lh_object_list["type"] == "table"]
    print("table list:")
    display(table_list)

    # create a temporary staging lakehouse per warehouse to create shortcuts into,
    # which point back to original warehouse data currently in the DR storage account
    lhname = f"temp_rlh_{i['primary_ws_name']}_{i['primary_name']}"
    lhname = lhname[:256]  # lakehouse name max length

    if p_logging_verbose:
        print(f'Checking whether the temporary lakehouse "{lhname}" exists in workspace {i["secondary_ws_name"]}...')

    temp_lh_id = getItemId(i['secondary_ws_id'], lhname, 'Lakehouse')

    if temp_lh_id == 'NotExists':
        payload = {
            "displayName": lhname,
            "description": (
                "Interim staging lakehouse for primary warehouse recovery: "
                f"{i['primary_ws_name']}_{i['primary_name']} into workspace "
                f"{i['secondary_ws_name']} ({i['secondary_ws_id']})"
            )
        }
        try:
            lhurl = f"v1/workspaces/{i['secondary_ws_id']}/lakehouses"
            lhresponse = fabric_post(lhurl, payload)
            temp_lh_id = lhresponse.json()['id']
            if p_logging_verbose:
                print(
                    f'Temporary lakehouse "{lhname}" created with Id {temp_lh_id}: '
                    f'{lhresponse.status_code} {lhresponse.text}'
                )
        except Exception as error:
            print(f"Error creating temp lakehouse {lhname}: {error}")
            continue
    else:
        if p_logging_verbose:
            print(f'Temporary lakehouse {lhname} ({temp_lh_id}) already exists.')

    # Create shortcuts for every table in the format of schema_table under the tables folder
    for _, itable in table_list.iterrows():
        shortcut_exists = False
        shortcut_name = f"{itable['schema']}_{itable['name']}"

        # Check if shortcut exists
        try:
            url = f"v1/workspaces/{i['secondary_ws_id']}/items/{temp_lh_id}/shortcuts/Tables/{shortcut_name}"
            tlhresponse = fabric_get(url)
            shortcut_exists = True
            if p_logging_verbose:
                print(f"Shortcut {shortcut_name} already exists")
        except Exception:
            shortcut_exists = False

        if not shortcut_exists:
            url = f"v1/workspaces/{i['secondary_ws_id']}/items/{temp_lh_id}/shortcuts"
            scpayload = {
                "path": "Tables/",
                "name": shortcut_name,
                "target": {
                    "oneLake": {
                        "workspaceId": i['primary_ws_id'],
                        "itemId": i['primary_id'],
                        "path": f"/Tables/{itable['schema']}/{itable['name']}"
                    }
                }
            }

            try:
                shctresponse = fabric_post(url, scpayload)
                if p_logging_verbose:
                    print(f"Shortcut {shortcut_name} created: {shctresponse.status_code}")
            except Exception as error:
                print(f"Error creating shortcut {shortcut_name}: {error}")

    recovery_pipeline_prefix = 'plRecover_WH6'
    recovery_pipeline = f"{recovery_pipeline_prefix}_{i['primary_ws_name']}_{i['primary_name']}"[:256]

    if p_logging_verbose:
        print('Attempting to deploy a copy pipeline in the target workspace to load the target warehouse tables from the shortcuts created above...')

    plid = getItemId(i['secondary_ws_id'], recovery_pipeline, 'DataPipeline')

    if plid == 'NotExists':
        plid = createDWrecoverypl(
            i['secondary_ws_id'],
            f"{recovery_pipeline_prefix}_{i['primary_ws_name']}_{i['primary_name']}"
        )
        if p_logging_verbose:
            print(f"Recovery pipeline {recovery_pipeline} created with Id {plid}")
    else:
        if p_logging_verbose:
            print(
                f'Datawarehouse recovery pipeline "{recovery_pipeline}" ({plid}) already exists in '
                f'workspace "{i["secondary_ws_name"]}" ({i["secondary_ws_id"]})'
            )
            print()

print('Done')  



StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 65, Finished, Available, Finished, False)

table name: Date
delta log in names
Date
table name: Geography
delta log in names
Geography
table name: HackneyLicense
delta log in names
HackneyLicense
table name: Medallion
delta log in names
Medallion
table name: Time
delta log in names
Time
table name: Trip
delta log in names
Trip
table name: Weather
delta log in names
Weather
dbo
lh_object_list:


SynapseWidget(Synapse.DataFrame, b4a31bcb-dcc2-4169-8bce-3ae7af8a313f)

schema list:


SynapseWidget(Synapse.DataFrame, 3841c186-5dc2-4be8-bf76-71b008eb3613)

table list:


SynapseWidget(Synapse.DataFrame, 4d130e6f-8b9d-43e8-b341-983b135654fb)

/home/trusted-service-user/cluster-env/trident_env/lib/python3.11/site-packages/sempy/fabric/_credentials.py:239: FutureWarning: The 'type' parameter is deprecated and will be removed in a future version. Use 'item_type' instead.
  return func(*args, **kwargs)
/home/trusted-service-user/cluster-env/trident_env/lib/python3.11/site-packages/sempy/fabric/_credentials.py:239: FutureWarning: The 'type' parameter is deprecated and will be removed in a future version. Use 'item_type' instead.
  return func(*args, **kwargs)


table name: Date
delta log in names
Date
table name: Geography
delta log in names
Geography
table name: HackneyLicense
delta log in names
HackneyLicense
table name: Medallion
delta log in names
Medallion
table name: Time
delta log in names
Time
table name: Trip
delta log in names
Trip
table name: Weather
delta log in names
Weather
dbo
lh_object_list:


SynapseWidget(Synapse.DataFrame, a1cd4559-04dc-4406-b1bb-36540f5df24c)

schema list:


SynapseWidget(Synapse.DataFrame, e51e5d02-a9cf-484e-8750-8a405f929680)

table list:


SynapseWidget(Synapse.DataFrame, d65a7ff5-8b43-4c52-a2ec-1acba0316c98)

/home/trusted-service-user/cluster-env/trident_env/lib/python3.11/site-packages/sempy/fabric/_credentials.py:239: FutureWarning: The 'type' parameter is deprecated and will be removed in a future version. Use 'item_type' instead.
  return func(*args, **kwargs)
/home/trusted-service-user/cluster-env/trident_env/lib/python3.11/site-packages/sempy/fabric/_credentials.py:239: FutureWarning: The 'type' parameter is deprecated and will be removed in a future version. Use 'item_type' instead.
  return func(*args, **kwargs)


Done


In [55]:
# wait for sql endpooint sync to reflect the new shortcuts that have just been created
time.sleep(60)

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 66, Finished, Available, Finished, False)

##### Stage 9: Recover warehouse data by running copy pipelines

In [56]:
import json
import time
import requests
from sempy.fabric import FabricRestClient

def get_token():
    return mssparkutils.credentials.getToken("pbi")

client = FabricRestClient(token_provider=get_token)

print("Starting warehouse recovery pipelines...")

for idx, i in enumerate(warehousedf):
    try:
        if p_logging_verbose:
            print(
                f"Invoking pipeline to copy warehouse data from "
                f"{i['primary_ws_name']}.{i['primary_name']} into "
                f"{i['secondary_ws_name']}.{i['secondary_name']}"
            )

        src_path = (
            f"abfss://{i['primary_ws_id']}@onelake.dfs.fabric.microsoft.com/"
            f"{i['primary_id']}"
        )

        lh_object_list = get_lh_object_list(src_path, ["Tables"])

        if lh_object_list is None or len(lh_object_list) == 0:
            print(
                f"Skipping {i['primary_ws_name']}.{i['primary_name']}: "
                f"no objects returned from source path {src_path}"
            )
            continue

        table_list = lh_object_list[lh_object_list["type"] == "table"].copy()

        if table_list.empty:
            print(
                f"Skipping {i['primary_ws_name']}.{i['primary_name']}: no tables found"
            )
            continue

        tables_to_copy = table_list[["schema", "name"]].to_dict(orient="records")

        lhname = f"temp_rlh_{i['primary_ws_name']}_{i['primary_name']}"
        temp_lh_id = getItemId(i["secondary_ws_id"], lhname, "Lakehouse")

        if not temp_lh_id or temp_lh_id == "NotExists":
            print(f"Skipping: temp lakehouse not found: {lhname}")
            continue

        lhurl = f"v1/workspaces/{i['secondary_ws_id']}/lakehouses/{temp_lh_id}"
        lhresponse = client.get(lhurl)

        if lhresponse.status_code != 200:
            print("Lakehouse GET status:", lhresponse.status_code)
            print("Lakehouse GET body:", lhresponse.text)
            continue

        lh_json = lhresponse.json()
        lhconnStr = lh_json["properties"]["sqlEndpointProperties"]["connectionString"]

        items = fabric.list_items(workspace=i["secondary_ws_id"])

        temp_lh_sqle = items[
            (items["Type"] == "SQLEndpoint") &
            (items["Display Name"] == lhname)
        ]

        if temp_lh_sqle.empty:
            print(f"Skipping: SQL endpoint not found for {lhname}")
            continue

        temp_lh_sqle_id = temp_lh_sqle["Id"].values[0]

        whurl = f"v1/workspaces/{i['secondary_ws_id']}/warehouses/{i['secondary_id']}"
        whresponse = client.get(whurl)

        if whresponse.status_code != 200:
            print("Warehouse GET status:", whresponse.status_code)
            print("Warehouse GET body:", whresponse.text)
            continue

        wh_json = whresponse.json()
        whconnStr = wh_json["properties"]["connectionInfo"]

        recovery_pipeline = (
            f"{recovery_pipeline_prefix}_{i['primary_ws_name']}_{i['primary_name'][:256]}"
        )

        plid = getItemId(i["secondary_ws_id"], recovery_pipeline, "DataPipeline")
        print("pipe id:", plid)

        if not plid or plid == "NotExists":
            print(f"Pipeline not found: {recovery_pipeline}")
            continue

        # Correct URL shape
        plurl = (
            f"https://api.fabric.microsoft.com/"
            f"v1/workspaces/{i['secondary_ws_id']}/items/{plid}/jobs/Pipeline/instances"
        )

        payload = {
            "executionData": {
                "parameters": {
                    "lakehouseId": temp_lh_sqle_id,
                    "tablesToCopy": tables_to_copy,
                    "workspaceId": i["secondary_ws_id"],
                    "warehouseId": i["secondary_id"],
                    "lakehouseConnStr": lhconnStr,
                    "warehouseConnStr": whconnStr
                }
            }
        }

        token = get_token()
        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json"
        }

        if p_logging_verbose:
            print("Submitting payload:")
            print(json.dumps(payload, indent=2))

        plresponse = requests.post(plurl, headers=headers, json=payload, timeout=120)

        if p_logging_verbose:
            print("POST status:", plresponse.status_code)
            print("POST headers:", dict(plresponse.headers))
            print("POST body:", plresponse.text)

        if plresponse.status_code not in (200, 201, 202):
            print(
                f"Pipeline submission failed for "
                f"{i['secondary_ws_name']}.{i['secondary_name']}"
            )
            continue

        if "Location" in plresponse.headers:
            print("Job location:", plresponse.headers["Location"])

        time.sleep(1)

    except Exception as e:
        print(
            f"Unhandled error while processing "
            f"{i.get('primary_ws_name', 'unknown')}.{i.get('primary_name', 'unknown')}: "
            f"{type(e).__name__}: {e}"
        )

print("Done")

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 67, Finished, Available, Finished, False)

Starting warehouse recovery pipelines...
table name: Date
delta log in names
Date
table name: Geography
delta log in names
Geography
table name: HackneyLicense
delta log in names
HackneyLicense
table name: Medallion
delta log in names
Medallion
table name: Time
delta log in names
Time
table name: Trip
delta log in names
Trip
pipe id: 66647916-106e-492b-b8e6-c81f8aff740b
Job location: https://api.fabric.microsoft.com/v1/workspaces/848c95f7-bd3b-426b-b89d-0192afae8a27/items/66647916-106e-492b-b8e6-c81f8aff740b/jobs/instances/64f5d569-577c-4231-957f-de7c2dbb90d1
pipe id: e57e9e22-8a9c-407e-90fb-1e9ebc23fd13
Job location: https://api.fabric.microsoft.com/v1/workspaces/13347c25-5de9-4613-9ef9-01526a6ed927/items/e57e9e22-8a9c-407e-90fb-1e9ebc23fd13/jobs/instances/98180a44-8c46-493f-a75f-c1a2bbefa3d1
Done


/home/trusted-service-user/cluster-env/trident_env/lib/python3.11/site-packages/sempy/fabric/_credentials.py:239: FutureWarning: The 'type' parameter is deprecated and will be removed in a future version. Use 'item_type' instead.
  return func(*args, **kwargs)
/home/trusted-service-user/cluster-env/trident_env/lib/python3.11/site-packages/sempy/fabric/_credentials.py:239: FutureWarning: The 'type' parameter is deprecated and will be removed in a future version. Use 'item_type' instead.
  return func(*args, **kwargs)


##### Stage 10: Add workspace roles assignments to the new workspaces

In [57]:
import json
from pyspark.sql import functions as F


print("Adding workspace role assignments...")

# Discover available lakehouse tables/schemas using utils helpers
discovered_tables = infer_lakehouse_tables()

if p_logging_verbose:
    print("Discovered lakehouse tables:")
    print(json.dumps(discovered_tables, indent=2))

rar_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="wsroleassignments_recovered",
    preferred_schema="dbo"
)

wks_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces",
    preferred_schema="dbo"
)

if p_logging_verbose:
    print("Using wsroleassignments path:", rar_path)
    print("Using workspaces path:", wks_path)

rar_df = spark.read.format("delta").load(rar_path)
wks_df = spark.read.format("delta").load(wks_path)

joined_df = (
    rar_df.alias("rar")
    .join(
        wks_df.alias("wks"),
        F.col("rar.workspaceName") == F.expr(
            f"replace(wks.Name, '{p_secondary_ws_suffix}', '')"
        ),
        "inner"
    )
    .filter(F.col("wks.name").endswith(p_secondary_ws_suffix))
    .filter(F.col("wks.id") != thisWsId)
    .select(
        F.col("wks.ID").alias("new_workspace_id"),
        F.col("wks.name").alias("new_workspace"),
        F.col("rar.*")
    )
)

# Only apply roles to selected workspaces
if len(p_list_of_workspaces_to_recover) > 0:
    joined_df = joined_df.filter(
        F.col("workspaceName").isin(p_list_of_workspaces_to_recover)
    )

# Ignore explicitly excluded workspaces
if len(p_ws_ignore_list) > 0:
    joined_df = joined_df.filter(
        ~F.col("workspaceName").isin(p_ws_ignore_list)
    )

# Ignore LIKE patterns
if len(p_ws_ignore_like_list) > 0:
    for notlike in p_ws_ignore_like_list:
        joined_df = joined_df.filter(
            ~F.col("new_workspace").like(notlike)
        )

dfwsrole = joined_df.collect()

for i in dfwsrole:
    ws_id = i["new_workspace_id"]
    ws_name = i["new_workspace"]

    wsroleurl = f"v1/workspaces/{ws_id}/roleAssignments"
    wsrolepayload = {
        "principal": {
            "id": i["principalId"],
            "type": i["principalType"]
        },
        "role": i["role"]
    }

    try:
        if p_logging_verbose:
            print(
                f"Attempting to add role assignment {i['role']} "
                f"for {i['principalType']} principal {i['displayName']} "
                f"({i['principalId']}) to workspace {ws_name} ({ws_id})..."
            )

        response = client.post(wsroleurl, json=wsrolepayload)

        if p_logging_verbose:
            print(
                f"Added role assignment successfully: "
                f"{response.status_code} {response.text}"
            )

    except Exception as error:
        error_text = str(error)
        parsed_message = None

        try:
            if "Error:" in error_text:
                error_part = error_text.split("Error:", 1)[1]
                if "Headers:" in error_part:
                    error_part = error_part.split("Headers:", 1)[0]
                error_json = json.loads(error_part.strip())
                parsed_message = error_json.get("message")
        except Exception:
            parsed_message = None

        if parsed_message:
            print(
                f"Couldn't add role assignment {i['role']} for principal "
                f"{i['displayName']} to workspace {ws_name} ({ws_id}). "
                f"Error: {parsed_message}"
            )
        else:
            print(
                f"Couldn't add role assignment {i['role']} for principal "
                f"{i['displayName']} to workspace {ws_name} ({ws_id}). "
                f"Raw error: {error_text}"
            )

print("Done")

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 68, Finished, Available, Finished, False)

Adding workspace role assignments...
Couldn't add role assignment Admin for principal System Administrator to workspace W1_DR (848c95f7-bd3b-426b-b89d-0192afae8a27). Error: The provided principal already has a role assigned in the workspace
Couldn't add role assignment Admin for principal System Administrator to workspace W2_DR (13347c25-5de9-4613-9ef9-01526a6ed927). Error: The provided principal already has a role assigned in the workspace
Done


##### Update default lakehouse for notebooks

Supports both standard and T-SQL notebooks

In [58]:
import json
from pyspark.sql import functions as F

print("Loading workspace/item mappings for notebook conversion...")

# Discover available lakehouse tables/schemas using utils helpers
discovered_tables = infer_lakehouse_tables()

if p_logging_verbose:
    print("Discovered lakehouse tables:")
    print(json.dumps(discovered_tables, indent=2))

# Resolve table paths
workspaces_recovered_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces_recovered",
    preferred_schema="dbo"
)

items_recovered_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="items_recovered",
    preferred_schema="dbo"
)

workspaces_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces",
    preferred_schema="dbo"
)

items_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="items",
    preferred_schema="dbo"
)

if p_logging_verbose:
    print("Using workspaces_recovered path:", workspaces_recovered_path)
    print("Using items_recovered path:", items_recovered_path)
    print("Using workspaces path:", workspaces_path)
    print("Using items path:", items_path)

# Load Delta tables
wr_df = spark.read.format("delta").load(workspaces_recovered_path)
ir_df = spark.read.format("delta").load(items_recovered_path)
wks_df = spark.read.format("delta").load(workspaces_path)
itm_df = spark.read.format("delta").load(items_path)


dfoldws = (
    wr_df.alias("wr")
    .join(
        wks_df.alias("wks"),
        F.col("wr.Name") == F.expr(f"replace(wks.Name, '{p_secondary_ws_suffix}', '')"),
        "inner"
    )
    .filter(F.col("wks.Name").endswith(p_secondary_ws_suffix))
    .filter(F.col("wks.Id") != thisWsId)
    .filter(F.col("wks.Name") != p_bcdr_workspace_src)
    .select(
        F.col("wr.Id").alias("old_ws_id"),
        F.col("wr.Name").alias("old_ws_name"),
        F.col("wks.Id").alias("new_ws_id"),
        F.col("wks.Name").alias("new_ws_name")
    )
)


dfolditms = (
    ir_df.alias("ir")
    .join(
        wr_df.alias("wr"),
        F.col("ir.workspaceId") == F.col("wr.Id"),
        "inner"
    )
    .join(
        wks_df.alias("wks"),
        F.col("wr.Name") == F.expr(f"replace(wks.Name, '{p_secondary_ws_suffix}', '')"),
        "inner"
    )
    .filter(F.col("wks.Name") != p_bcdr_workspace_src)
    .filter(F.col("wks.Name").endswith(p_secondary_ws_suffix))
    .filter(F.col("wks.Id") != thisWsId)
    .select(
        F.col("ir.Id").alias("old_item_id"),
        F.col("ir.displayName").alias("old_item_name"),
        F.col("wr.Id").alias("old_ws_id"),
        F.col("wr.Name").alias("old_ws_name"),
        F.col("wks.Id").alias("new_ws_id"),
        F.col("wks.Name").alias("new_ws_name")
    )
)

dfwsnb_df = (
    itm_df.alias("itm")
    .join(
        wks_df.alias("wks"),
        F.col("itm.workspaceId") == F.col("wks.Id"),
        "inner"
    )
    .filter(F.col("wks.Name") != p_bcdr_workspace_src)
    .filter(F.col("itm.type") == "Notebook")
    .filter(F.col("wks.Name").endswith(p_secondary_ws_suffix))
    .filter(F.col("wks.Id") != thisWsId)
    .select(
        F.col("wks.Id").alias("new_workspace_id"),
        F.col("wks.Name").alias("new_workspace_name"),
        F.col("itm.*")
    )
)

# Ignore workspaces explicitly listed
if len(p_ws_ignore_list) > 0:
    dfwsnb_df = dfwsnb_df.filter(
        ~F.col("new_workspace_name").isin(p_ws_ignore_list)
    )

# Ignore LIKE patterns
if len(p_ws_ignore_like_list) > 0:
    for notlike in p_ws_ignore_like_list:
        dfwsnb_df = dfwsnb_df.filter(
            ~F.col("new_workspace_name").like(f"%{notlike}%")
        )

print("Looping through notebooks...")

dfwsnb = dfwsnb_df.collect()

for idx, i in enumerate(dfwsnb):
    print(f"Updating notebook {i['displayName']}")

    # Get the current notebook definition
    notebook_def = notebookutils.notebook.getDefinition(
        i["displayName"],
        workspaceId=i["new_workspace_id"]
    )

    json_payload = json.loads(notebook_def)

    # update default lakehouse if exists
    if (
        "metadata" in json_payload
        and "dependencies" in json_payload["metadata"]
        and "lakehouse" in json_payload["metadata"]["dependencies"]
        and json_payload["metadata"]["dependencies"]["lakehouse"] is not None
    ):
        current_lakehouse = json_payload["metadata"]["dependencies"]["lakehouse"]

        if "default_lakehouse_name" in current_lakehouse:
            old_lakehouse_id = current_lakehouse["default_lakehouse"]

            old_lakehouse_match = (
                dfolditms
                .filter(F.col("old_item_id") == old_lakehouse_id)
                .select(
                    "old_ws_id",
                    "old_ws_name",
                    "new_ws_id",
                    "new_ws_name"
                )
                .collect()
            )

            if not old_lakehouse_match:
                print(
                    f"Could not find DR mapping for default lakehouse "
                    f"{current_lakehouse['default_lakehouse_name']} "
                    f"in notebook {i['displayName']}, skipping."
                )
                continue

            old_lakehouse_ws_id, old_lakehouse_ws_name, new_lakehouse_ws_id, new_lakehouse_ws_name = \
                old_lakehouse_match[0]

            print(
                "Converting notebook " + i["displayName"] +
                " in workspace " + i["new_workspace_name"] + " (" + i["new_workspace_id"] + ")" +
                " with default lakehouse " + current_lakehouse["default_lakehouse_name"] +
                " in previous workspace " + old_lakehouse_ws_name + " (" + old_lakehouse_ws_id + ")" +
                " to " + new_lakehouse_ws_name + " (" + new_lakehouse_ws_id + ")"
            )

            json_payload["metadata"]["dependencies"]["lakehouse"] = {}

            notebookutils.notebook.updateDefinition(
                name=i["displayName"],
                content=json.dumps(json_payload),
                defaultLakehouse=current_lakehouse["default_lakehouse_name"],
                defaultLakehouseWorkspace=new_lakehouse_ws_id,
                workspaceId=i["new_workspace_id"]
            )

            print(f"Updated notebook {i['displayName']}")

        else:
            print(f"No default lakehouse was found in the source notebook {i['displayName']}")

    # update default warehouse if exists
    elif (
        "metadata" in json_payload
        and "dependencies" in json_payload["metadata"]
        and "warehouse" in json_payload["metadata"]["dependencies"]
        and json_payload["metadata"]["dependencies"]["warehouse"] is not None
    ):
        current_warehouse = json_payload["metadata"]["dependencies"]["warehouse"]
        old_warehouse_id = current_warehouse["default_warehouse"]

        old_warehouse_match = (
            dfolditms
            .filter(F.col("old_item_id") == old_warehouse_id)
            .select(
                "old_item_name",
                "old_ws_id",
                "old_ws_name",
                "new_ws_id",
                "new_ws_name"
            )
            .collect()
        )

        if not old_warehouse_match:
            print(
                f"Could not find DR mapping for warehouse id {old_warehouse_id} "
                f"in notebook {i['displayName']}, skipping."
            )
            continue

        old_wh_name, old_warehouse_ws_id, old_warehouse_ws_name, new_warehouse_ws_id, new_warehouse_ws_name = \
            old_warehouse_match[0]

        target_wh_id = fabric.resolve_item_id(
            item_name=old_wh_name,
            type="Warehouse",
            workspace=new_warehouse_ws_id
        )

        if "default_warehouse" in current_warehouse:
            print(
                "Converting notebook " + i["displayName"] +
                " in workspace " + i["new_workspace_name"] + " (" + i["new_workspace_id"] + ")" +
                " with default warehouse " + old_wh_name +
                " in previous workspace " + old_warehouse_ws_name + " (" + old_warehouse_ws_id + ")" +
                " to " + old_wh_name + " in new workspace " +
                new_warehouse_ws_name + " (" + new_warehouse_ws_id + ")"
            )

            json_payload["metadata"]["dependencies"]["warehouse"]["default_warehouse"] = target_wh_id

            for warehouse in json_payload["metadata"]["dependencies"]["warehouse"]["known_warehouses"]:
                if warehouse["id"] == old_warehouse_id:
                    warehouse["id"] = target_wh_id

            notebookutils.notebook.updateDefinition(
                name=i["displayName"],
                content=json.dumps(json_payload),
                workspaceId=i["new_workspace_id"]
            )

            print(f"Updated notebook {i['displayName']}")

        else:
            print(f"No default warehouse was found in the source notebook {i['displayName']}")

    else:
        print(f"No default lakehouse/warehouse set for notebook {i['displayName']}, ignoring.")

print("Done")

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 69, Finished, Available, Finished, False)

Loading workspace/item mappings for notebook conversion...
Looping through notebooks...
Done


##### Update direct lake model lakehouse/warehouse connection
    

In [60]:
from pyspark.sql import functions as F

print("Loading semantic models...")

# Discover available lakehouse tables/schemas using utils helpers
discovered_tables = infer_lakehouse_tables()

if p_logging_verbose:
    print("Discovered lakehouse tables:")
    print(json.dumps(discovered_tables, indent=2))

items_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="items",
    preferred_schema="dbo"
)

workspaces_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces",
    preferred_schema="dbo"
)

if p_logging_verbose:
    print("Using items path:", items_path)
    print("Using workspaces path:", workspaces_path)

itm_df = spark.read.format("delta").load(items_path)
wks_df = spark.read.format("delta").load(workspaces_path)

dfwssm_df = (
    itm_df.alias("itm")
    .join(
        wks_df.alias("wks"),
        F.col("itm.workspaceId") == F.col("wks.Id"),
        "inner"
    )
    .filter(F.col("itm.type") == "SemanticModel")
    .filter(F.col("wks.Name").endswith(p_secondary_ws_suffix))
    .filter(F.col("wks.Id") != thisWsId)
    .select(
        F.col("wks.Id").alias("new_workspace_id"),
        F.col("wks.Name").alias("new_workspace_name"),
        F.col("itm.*")
    )
)

# Ignore workspaces based on explicit ignore list
if len(p_ws_ignore_list) > 0:
    dfwssm_df = dfwssm_df.filter(
        ~F.col("new_workspace_name").isin(p_ws_ignore_list)
    )

# Ignore workspaces based on LIKE patterns
if len(p_ws_ignore_like_list) > 0:
    for notlike in p_ws_ignore_like_list:
        dfwssm_df = dfwssm_df.filter(
            ~F.col("new_workspace_name").like(f"%{notlike}%")
        )

dfwssm = dfwssm_df.collect()

# Iterate over each semantic model
for idx, row in enumerate(dfwssm):

    # Check if the dataset is not the default semantic model
    if not labs.is_default_semantic_model(row["displayName"], row["new_workspace_id"]):
        print(
            "Updating semantic model connection "
            + row["displayName"]
            + " in workspace "
            + row["new_workspace_name"]
        )

        old_ws_name = row["new_workspace_name"].replace(p_secondary_ws_suffix, "")

        source_type, source_name = get_direct_lake_source(
            row["displayName"],
            workspace=old_ws_name
        )

        labs.directlake.update_direct_lake_model_connection(
            dataset=row["displayName"],
            workspace=row["new_workspace_name"],
            source=source_name,
            source_type=source_type,
            source_workspace=row["new_workspace_name"]
        )

        labs.refresh_semantic_model(
            dataset=row["displayName"],
            workspace=row["new_workspace_name"]
        )

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 71, Finished, Available, Finished, False)

Loading semantic models...
Updating semantic model connection SemanticModel1 in workspace W1_DR


ValueError: 🔴 SQL Endpoint not found.

In [62]:
print(row["displayName"], row["new_workspace_name"], source_type, source_name)

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 73, Finished, Available, Finished, False)

NameError: name 'source_type' is not defined

##### Update direct lake model lakehouse/warehouse connection

In [66]:
import json
from pyspark.sql import functions as F

print("Loading semantic models...")

discovered_tables = infer_lakehouse_tables()

items_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="items",
    preferred_schema=None
)

workspaces_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces",
    preferred_schema=None
)

itm_df = spark.read.format("delta").load(items_path)
wks_df = spark.read.format("delta").load(workspaces_path)

dfwssm_df = (
    itm_df.alias("itm")
    .join(
        wks_df.alias("wks"),
        F.col("itm.workspaceId") == F.col("wks.Id"),
        "inner"
    )
    .filter(F.col("itm.type") == "SemanticModel")
    .filter(F.col("wks.Name").endswith(p_secondary_ws_suffix))
    .filter(F.col("wks.Id") != thisWsId)
    .select(
        F.col("wks.Id").alias("new_workspace_id"),
        F.col("wks.Name").alias("new_workspace_name"),
        F.col("itm.*")
    )
)

if len(p_ws_ignore_list) > 0:
    dfwssm_df = dfwssm_df.filter(
        ~F.col("new_workspace_name").isin(p_ws_ignore_list)
    )

if len(p_ws_ignore_like_list) > 0:
    for notlike in p_ws_ignore_like_list:
        dfwssm_df = dfwssm_df.filter(
            ~F.col("new_workspace_name").like(f"%{notlike}%")
        )

dfwssm = dfwssm_df.collect()

for row in dfwssm:
    if labs.is_default_semantic_model(row["displayName"], row["new_workspace_id"]):
        continue

    print(
        f"Updating semantic model connection {row['displayName']} "
        f"in workspace {row['new_workspace_name']}"
    )

    old_ws_name = row["new_workspace_name"].replace(p_secondary_ws_suffix, "")

    sources = labs.directlake.get_direct_lake_sources(
    dataset=row["displayName"],
    workspace=old_ws_name
    )

    if not sources:
        print(f"Skipping {row['displayName']} - no Direct Lake source found in workspace {old_ws_name}")
        continue

        src = sources[0]  # assuming one source
        source_type = src["itemType"]       # "Lakehouse" or "Warehouse"
        source_name = src["itemName"]       # source artifact name
        source_workspace = src["workspaceName"]
        uses_sql_endpoint = src.get("usesSqlEndpoint")

        print(
            f"Resolved source for {row['displayName']}: "
            f"type={source_type}, name={source_name}, workspace={source_workspace}, "
            f"uses_sql_endpoint={uses_sql_endpoint}"
        )

        labs.directlake.update_direct_lake_model_connection(
            dataset=row["displayName"],
            workspace=row["new_workspace_name"],
            source=source_name,
            source_type=source_type,
            source_workspace=row["new_workspace_name"],
            use_sql_endpoint=False
        )
        labs.refresh_semantic_model(
            dataset=row["displayName"],
            workspace=row["new_workspace_name"]
        )

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 77, Finished, Available, Finished, False)

Loading semantic models...
Updating semantic model connection SemanticModel1 in workspace W1_DR
Updating semantic model connection SemanticModel2 in workspace W2_DR


##### Update data pipeline source & sink connections

Support changes lakehouses, warehouses, notebooks and connections from source to target. <br>
Connections changes should be expressed as an array of tuples [(from_1:to_1),)from_N:to_N)]

Example usage:
    p_new_json = swap_pipeline_connection(pipeline_json,dfwsitms,
            ['Warehouse','Lakehouse','Notebook'],
            [('CONNECTION_ID_FROM','CONNECTION_ID_TO'),('CONNECTION_ID_FROM','CONNECTION_NAME_TO'),('CONNECTION_NAME_FROM','CONNECTION_ID_TO')]) 


In [67]:
import json
from pyspark.sql import functions as F

connections_from_to = [
    ('1b478585-f04d-49bf-b31d-cd57c3d8ca25', '11dd2898-018a-4d52-843a-3cac828984c4')
]

print("Loading pipeline dependency items...")

discovered_tables = infer_lakehouse_tables()

items_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="items",
    preferred_schema=None
)

items_recovered_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="items_recovered",
    preferred_schema=None
)

workspaces_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces",
    preferred_schema=None
)

workspaces_recovered_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces_recovered",
    preferred_schema=None
)

itm_df = spark.read.format("delta").load(items_path)
itm_rec_df = spark.read.format("delta").load(items_recovered_path)
wks_df = spark.read.format("delta").load(workspaces_path)
wks_rec_df = spark.read.format("delta").load(workspaces_recovered_path)

# Optional: makes the replace() join easier to read/debug
wks_base_df = wks_df.withColumn(
    "base_workspace_name",
    F.expr(f"replace(Name, '{p_secondary_ws_suffix}', '')")
)

dfwsitms_df = (
    itm_df.alias("itm")
    .join(
        wks_base_df.alias("wks"),
        F.col("itm.workspaceId") == F.col("wks.Id"),
        "inner"
    )
    .join(
        itm_rec_df.alias("itm_rec"),
        (F.col("itm.displayName") == F.col("itm_rec.displayName")) &
        (F.col("itm.type") == F.col("itm_rec.type")),
        "inner"
    )
    .join(
        wks_rec_df.alias("wks_rec"),
        (F.col("wks_rec.Name") == F.col("wks.base_workspace_name")) &
        (F.col("wks.Id") != F.lit(thisWsId)) &
        (F.col("wks_rec.Id") == F.col("itm_rec.workspaceId")),
        "inner"
    )
    .filter(F.col("itm_rec.Id") != F.col("itm.Id"))
    .filter(F.col("wks.Name") != F.col("wks_rec.Name"))
    .select(
        F.col("itm.type").alias("type"),
        F.col("itm_rec.Id").alias("old_item_id"),
        F.col("itm.Id").alias("new_item_id"),
        F.col("itm.displayName").alias("itm_name"),
        F.col("wks_rec.Id").alias("old_ws_id"),
        F.col("wks_rec.Name").alias("old_ws_name"),
        F.col("wks.Id").alias("new_ws_id"),
        F.col("wks.Name").alias("new_ws_name")
    )
)

# Ignore workspaces based on the ignore list
if len(p_ws_ignore_list) > 0:
    dfwsitms_df = dfwsitms_df.filter(
        ~F.col("new_ws_name").isin(p_ws_ignore_list)
    )

if len(p_ws_ignore_like_list) > 0:
    for notlike in p_ws_ignore_like_list:
        dfwsitms_df = dfwsitms_df.filter(
            ~F.col("new_ws_name").like(f"%{notlike}%")
        )

all_items = dfwsitms_df.collect()
dfwspl = [r for r in all_items if r["type"] == "DataPipeline"]

print("Looping through pipelines...")

for i in dfwspl:
    pipeline_name = i["itm_name"]
    workspace_name = i["new_ws_name"]

    print(f"Processing pipeline {pipeline_name} in {workspace_name}")

    pipeline_json = json.loads(
        labs.get_data_pipeline_definition(pipeline_name, workspace_name)
    )

    pipeline_json_bak = json.dumps(pipeline_json, sort_keys=True)

    p_new_json = swap_pipeline_connection_paths(
        pipeline_json=pipeline_json,
        rows=all_items,
        discovered_tables=discovered_tables,
        replace_types=["Warehouse", "Lakehouse", "Notebook"]
    )[0]

    p_new_json_str = json.dumps(p_new_json, sort_keys=True)

    if pipeline_json_bak != p_new_json_str:
        update_data_pipeline_definition(
            name=pipeline_name,
            pipeline_content=p_new_json,
            workspace=workspace_name
        )
        print(f"Updated pipeline {pipeline_name}")
    else:
        print(f"No changes made for pipeline {pipeline_name}")

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 78, Finished, Available, Finished, False)

Loading pipeline dependency items...
Looping through pipelines...


##### Rebind reports in new DR workspace

https://semantic-link-labs.readthedocs.io/en/latest/sempy_labs.report.html#sempy_labs.report.report_rebind

In [68]:
from pyspark.sql import functions as F

print("Loading report/dataset mappings...")

# Discover available lakehouse tables/schemas
discovered_tables = infer_lakehouse_tables()

if p_logging_verbose:
    print("Discovered lakehouse tables:")
    print(json.dumps(discovered_tables, indent=2))

items_recovered_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="items_recovered",
    preferred_schema=None
)

workspaces_recovered_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces_recovered",
    preferred_schema=None
)

items_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="items",
    preferred_schema=None
)

workspaces_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="workspaces",
    preferred_schema=None
)

reports_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="reports",
    preferred_schema=None
)

reports_recovered_path = find_table_path(
    discovered_tables=discovered_tables,
    table_name="reports_recovered",
    preferred_schema=None
)

if p_logging_verbose:
    print("Using items_recovered path:", items_recovered_path)
    print("Using workspaces_recovered path:", workspaces_recovered_path)
    print("Using items path:", items_path)
    print("Using workspaces path:", workspaces_path)
    print("Using reports path:", reports_path)
    print("Using reports_recovered path:", reports_recovered_path)

ir_df = spark.read.format("delta").load(items_recovered_path)
wr_df = spark.read.format("delta").load(workspaces_recovered_path)
itm_df = spark.read.format("delta").load(items_path)
wks_df = spark.read.format("delta").load(workspaces_path)
rep_df = spark.read.format("delta").load(reports_path)
rep_rec_df = spark.read.format("delta").load(reports_recovered_path)


dfolditms = (
    ir_df.alias("ir")
    .join(
        wr_df.alias("wr"),
        F.col("ir.workspaceId") == F.col("wr.Id"),
        "inner"
    )
    .join(
        wks_df.alias("wks"),
        F.col("wr.Name") == F.expr(f"replace(wks.Name, '{p_secondary_ws_suffix}', '')"),
        "inner"
    )
    .filter(F.col("wks.Name").endswith(p_secondary_ws_suffix))
    .filter(F.col("wks.Id") != thisWsId)
    .select(
        F.col("ir.Id").alias("old_item_id"),
        F.col("ir.displayName").alias("old_item_name"),
        F.col("wr.Id").alias("old_ws_id"),
        F.col("wr.Name").alias("old_ws_name"),
        F.col("wks.Id").alias("new_ws_id"),
        F.col("wks.Name").alias("new_ws_name")
    )
)


dfnewitms = (
    itm_df.alias("itm")
    .join(
        wks_df.alias("wks"),
        F.col("itm.workspaceId") == F.col("wks.Id"),
        "inner"
    )
    .filter(F.col("wks.Name").endswith(p_secondary_ws_suffix))
    .filter(F.col("wks.Id") != thisWsId)
    .select(
        F.col("itm.Id").alias("new_item_id"),
        F.col("itm.displayName").alias("new_item_name"),
        F.col("wks.Id").alias("new_ws_id"),
        F.col("wks.Name").alias("new_ws_name")
    )
)


dfwssm_df = (
    rep_df.alias("rep")
    .join(
        itm_df.alias("itm"),
        F.col("rep.Id") == F.col("itm.Id"),
        "inner"
    )
    .join(
        wks_df.alias("wks"),
        F.col("itm.workspaceId") == F.col("wks.Id"),
        "inner"
    )
    .join(
        rep_rec_df.alias("rep_rec"),
        F.col("rep.Name") == F.col("rep_rec.Name"),
        "inner"
    )
    .join(
        ir_df.alias("itm_rec"),
        F.col("itm_rec.Id") == F.col("rep_rec.Id"),
        "inner"
    )
    .join(
        wr_df.alias("wks_rec"),
        (F.col("wks_rec.Name") == F.expr(f"replace(wks.Name, '{p_secondary_ws_suffix}', '')")) &
        (F.col("wks_rec.Id") == F.col("itm_rec.workspaceId")),
        "inner"
    )
    .filter(F.col("rep_rec.Id") != F.col("rep.Id"))
    .filter(F.col("wks.Name") != F.col("wks_rec.Name"))
    .select(
        F.col("rep.Name").alias("Name"),
        F.col("rep.Dataset_Id").alias("old_dataset_id"),
        F.col("rep_rec.Id").alias("old_rep_id"),
        F.col("wks_rec.Id").alias("old_ws_id"),
        F.col("wks_rec.Name").alias("old_ws_name"),
        F.col("wks.Id").alias("new_ws_id"),
        F.col("wks.Name").alias("new_ws_name")
    )
)

# Ignore workspaces based on explicit ignore list
if len(p_ws_ignore_list) > 0:
    dfwssm_df = dfwssm_df.filter(
        ~F.col("new_ws_name").isin(p_ws_ignore_list)
    )

# Ignore LIKE patterns
if len(p_ws_ignore_like_list) > 0:
    for notlike in p_ws_ignore_like_list:
        dfwssm_df = dfwssm_df.filter(
            ~F.col("new_ws_name").like(f"%{notlike}%")
        )

dfwssm = dfwssm_df.orderBy("new_ws_id").collect()

current_workspace_id = ""

# Iterate over each report in the dataframe
for idx, row in enumerate(dfwssm):
    old_dataset_matches = (
        dfolditms
        .filter(F.col("old_item_id") == row["old_dataset_id"])
        .select(
            "old_item_name",
            "old_ws_id",
            "old_ws_name",
            "new_ws_id",
            "new_ws_name"
        )
        .collect()
    )

    if old_dataset_matches:
        old_dataset_name, old_dataset_ws_id, old_dataset_ws_name, new_dataset_ws_id, new_dataset_ws_name = old_dataset_matches[0]

        print(
            f"Rebinding report {row['Name']} in workspace {row['new_ws_name']} "
            f"with dataset name {old_dataset_name} in target workspace {new_dataset_ws_name}"
        )

        labs.report.report_rebind(
            report=row["Name"],
            dataset=old_dataset_name,
            report_workspace=row["new_ws_name"],
            dataset_workspace=new_dataset_ws_name
        )
    else:
        new_dataset_matches = (
            dfnewitms
            .filter(F.col("new_item_id") == row["old_dataset_id"])
            .select(
                "new_item_name",
                "new_ws_id",
                "new_ws_name"
            )
            .collect()
        )

        if new_dataset_matches:
            new_item_name, new_ws_id, new_ws_name = new_dataset_matches[0]
            print(
                f"Report {row['Name']} in workspace {row['new_ws_name']} "
                f"has a dataset {new_item_name} in a workspace {new_ws_name} ({new_ws_id})"
            )
        else:
            print(
                f"Could not resolve dataset mapping for report {row['Name']} "
                f"in workspace {row['new_ws_name']} (old_dataset_id={row['old_dataset_id']})"
            )

StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, 79, Finished, Available, Finished, False)

Loading report/dataset mappings...
Report Passenger Trends in workspace W2_DR has a dataset SemanticModel2 in a workspace W2_DR (13347c25-5de9-4613-9ef9-01526a6ed927)
Report Product Profitability in workspace W1_DR has a dataset SemanticModel1 in a workspace W1_DR (848c95f7-bd3b-426b-b89d-0192afae8a27)


##### Clean up - deletes recovered workspaces!!
<div style="display: flex; align-items: flex-end;"><img style="float: left; margin-right: 10px;" src="https://github.com/hurtn/images/blob/main/stop.png?raw=true" width="50"><span><h6>Only run the cell below if you are re-testing this process  or do not wish to keep the recovered workspaces in the secondary.<br>Keeping the cell frozen ensures it is not run when the Run all button is used.</span></div>

In [ ]:
# Please ensure you have run the workspaceutils command at the top of this notebook before running this cell to ensure all necessary imports and variables are loaded.

print('Refreshing the workspaces metadata table...')
# Refresh the list of current workspaces
saveWorkspaceMeta()

thisWsId = notebookutils.runtime.context['currentWorkspaceId'] #obtaining this so we don't accidently delete this workspace!

delete_ws_sql = "SELECT distinct ID,Type,Name FROM workspaces where Name like '%"+p_secondary_ws_suffix+"' and id != '"+thisWsId+"' and Name != '" +p_bcdr_workspace_src+"'" 

# Ingore workspaces based on the ignore list defined in the parameter section at the top of this notebook
if len(p_ws_ignore_list)>0:
    delete_ws_sql = delete_ws_sql+ " and name not in ('" + "', '".join(p_ws_ignore_list)+ "') "

if len(p_ws_ignore_like_list)>0:
    for notlike in p_ws_ignore_like_list:
        delete_ws_sql  = delete_ws_sql + " and name not like '" + notlike + "'"


print('Deleting workspaces...')
# Get all workspaces created with the prefix set in the parameters at the top so that they can be deleted, except for this workspace of course!
df = spark.sql(delete_ws_sql).collect()
for i in df:
    #print(i['ID'])
    if i['Type'] == 'Workspace':
      workspaceId = i['ID']
      if p_logging_verbose:
        print("Deleting workspace "+i['Name'] + '(' + i['ID'] + ')')
      response = client.delete("/v1/workspaces/"+workspaceId)
      if p_logging_verbose and response.status_code ==200:
        print('Successfully deleted')

print('Refreshing the workspaces metadata table after deleting recovered workspaces...')
# now refresh workspaces
saveWorkspaceMeta()
print('Done')


StatementMeta(, f0d69e60-a479-416a-85d2-ec75ccc7905c, -1, Cancelled, , Cancelled, True)